typescript/lancedb-rag/index.ts

```diff
@@ -0,0 +1,196 @@
+import "dotenv/config";
+
+import fs from "node:fs";
+import readline from "node:readline/promises";
+import { OpenAIEmbeddingFunction, connect, type Connection } from "vectordb";
+import OpenAI from "openai";
+
+import { LiteralClient } from "@literalai/client";
+
+const DATA_URL = `https://datasets-server.huggingface.co/rows?dataset=jamescalam%2Fyoutube-transcriptions&config=default&split=train&offset=0&length=100`;
+
+const RAW_DATA_PATH = "data/raw-data.json";
+
+type RawRow = {
+  id: string;
+  title: string;
+  published: string;
+  url: string;
+  video_id: string;
+  channel_id: string;
+  text: string;
+  start: number;
+  end: number;
+};
+
+/**
+ * Get raw data
+ * Download them if from the Hugging Face Datasets server
+ * Then store locally for latter use.
+ */
+const getRawData = async (): Promise<{ rows: { row: RawRow }[] }> => {
+  if (fs.existsSync(RAW_DATA_PATH)) {
+    const rawData = fs.readFileSync(RAW_DATA_PATH, "utf8");
+    return JSON.parse(rawData);
+  }
+  const response = await fetch(DATA_URL);
+  const content = await response.text();
+  fs.writeFileSync(RAW_DATA_PATH, content, "utf8");
+  return JSON.parse(content);
+};
+
+/**
+ * Create the embeddings table
+ */
+const createEmbeddingsTable = async (
+  db: Connection,
+  embedFunction: OpenAIEmbeddingFunction
+) => {
+  // read the input file into a JSON array
+  const rawData = await getRawData();
+  const lines = rawData.rows.map((r) => r.row);
+  const data = contextualize(lines, 20, "video_id");
+  return await db.createTable("vectors", data, embedFunction);
+};
+
+/**
+ * Each transcript has a small text column, we include previous transcripts in order to
+ * have more context information when creating embeddings
+ */
+const contextualize = (
+  rows: RawRow[],
+  contextSize: number,
+  groupColumn: keyof RawRow
+) => {
+  const grouped: Record<string, RawRow[]> = {};
+  rows.forEach((row) => {
+    if (!grouped[row[groupColumn]]) {
+      grouped[row[groupColumn]] = [];
+    }
+    grouped[row[groupColumn]].push(row);
+  });
+
+  const data: (RawRow & { context: string })[] = [];
+  Object.keys(grouped).forEach((key) => {
+    data.push(
```

Comment:
in general, `forEach => push` can be simplified as a `.map`

Reply:
(or maybe `flatMap` in this case)

Reply:
done